File: packages/tutoriales/src/centro/tarjeta-de-tutorial.ts
import { ChangeDetectionStrategy, Component, computed, input, output } from '@angular/core'
import { Boton } from '@aportaya/ui/boton/boton'
import { ChipEstado } from '@aportaya/ui/chip-estado/chip-estado'
import { Progreso } from '@aportaya/ui/progreso/progreso'
import type { Tono } from '@aportaya/ui/tono/tono'
import type { TutorialEnLista } from './vista-de-tutoriales'
import { textosAyuda } from './textos'

/** Un tutorial en la lista: qué enseña, cuánto lleva y qué botón corresponde. */
@Component({
  selector: 'ap-tarjeta-de-tutorial',
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [Boton, ChipEstado, Progreso],
  template: `
    <header>
      <div class="titulo">
        <h3>{{ fila().tutorial.titulo }}</h3>
        <p class="meta">
          {{ fila().tutorial.categoria }} · {{ t.dificultad[fila().tutorial.dificultad] }} · {{ t.pasos(fila().tutorial.pasos.length) }}
          @if (fila().tutorial.minutos; as m) { · {{ t.minutos(m) }} }
        </p>
      </div>
      <div class="chips">
        @if (fila().tutorial.obligatorio) {
          <ap-chip-estado tono="aviso">{{ t.obligatorio }}</ap-chip-estado>
        }
        <ap-chip-estado [tono]="tono()">{{ t.estado[fila().estado] }}</ap-chip-estado>
      </div>
    </header>
    <p class="descripcion">{{ fila().tutorial.descripcion }}</p>

    @if (fila().requisitosPendientes.length > 0) {
      <p class="requisitos">{{ t.requisitos }}: {{ nombresDeRequisitos() }}</p>
    }
    @if (fila().fraccion > 0 && fila().estado !== 'completado') {
      <ap-progreso [valor]="fila().fraccion" [etiqueta]="'Avance de ' + fila().tutorial.titulo" />
    }

    <div class="acciones">
      @if (fila().continuable) {
        <ap-boton variante="primario" tamano="sm" (pulsado)="continuar.emit()">{{ t.continuar }}</ap-boton>
      } @else {
        <ap-boton variante="primario" tamano="sm" (pulsado)="comenzar.emit()">{{ fila().estado === 'completado' ? t.repetir : t.comenzar }}</ap-boton>
      }
      @if (fila().progreso !== undefined) {
        <ap-boton variante="fantasma" tamano="sm" (pulsado)="reiniciar.emit()">{{ t.reiniciar }}</ap-boton>
      }
    </div>
  `,
  styles: `
    :host { display: flex; flex-direction: column; gap: var(--s3); padding: var(--s4); border: var(--borde-fino) solid var(--border); border-radius: var(--r-lg); background: var(--surface); }
    header { display: flex; justify-content: space-between; align-items: flex-start; gap: var(--s3); flex-wrap: wrap; }
    h3 { margin: 0; font: var(--t-titulo-3); letter-spacing: var(--t-titulo-3-track); }
    .meta { margin: var(--s1) 0 0; color: var(--text-3); font: var(--t-cuerpo-chico); }
    .chips { display: flex; gap: var(--s2); flex-wrap: wrap; }
    .descripcion { margin: 0; color: var(--text-2); font: var(--t-cuerpo); }
    .requisitos { margin: 0; color: var(--aviso-texto); font: var(--t-cuerpo-chico); }
    .acciones { display: flex; gap: var(--s2); flex-wrap: wrap; margin-top: auto; }
  `,
})
export class TarjetaDeTutorial {
  protected readonly t = textosAyuda
  readonly fila = input.required<TutorialEnLista>()
  readonly comenzar = output<void>()
  readonly continuar = output<void>()
  readonly reiniciar = output<void>()

  /** El estado también se dice con palabra, no solo con color (regla de accesibilidad). */
  protected readonly tono = computed<Tono>(() => {
    switch (this.fila().estado) {
      case 'completado':
        return 'ok'
      case 'en-progreso':
        return 'info'
      case 'omitido':
        return 'aviso'
      default:
        return 'neutro'
    }
  })

  protected readonly nombresDeRequisitos = computed(() => this.fila().requisitosPendientes.map((r) => r.titulo).join(', '))
}

File: apps/movil/lib/pantallas/soporte/tarjeta_de_tutorial.dart
import 'package:aportaya_diseno/atomos/boton.dart';
import 'package:aportaya_diseno/atomos/boton_tamano.dart';
import 'package:aportaya_diseno/atomos/boton_variante.dart';
import 'package:aportaya_diseno/atomos/chip_estado.dart';
import 'package:aportaya_diseno/atomos/progreso.dart';
import 'package:aportaya_diseno/atomos/tono.dart';
import 'package:aportaya_diseno/moleculas/tarjeta.dart';
import 'package:aportaya_diseno/tokens/tokens.dart';
import 'package:flutter/material.dart';

import '../../dominio/tutoriales/modelo.dart';
import 'textos.dart';
import 'vista_de_tutoriales.dart';

/// Un tutorial en la lista: qué enseña, cuánto lleva y qué botón corresponde.
class TarjetaDeTutorial extends StatelessWidget {
  const TarjetaDeTutorial({
    super.key,
    required this.fila,
    required this.alComenzar,
    required this.alContinuar,
    required this.alReiniciar,
  });

  final TutorialEnLista fila;
  final VoidCallback alComenzar;
  final VoidCallback alContinuar;
  final VoidCallback alReiniciar;

  @override
  Widget build(BuildContext context) {
    final t = Tokens.of(context);
    final tutorial = fila.tutorial;
    return Tarjeta(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Expanded(
                child: Text(
                  tutorial.titulo,
                  style: Tipo.titulo3.copyWith(color: t.text),
                ),
              ),
              const SizedBox(width: Espacio.s2),
              // El estado se dice con palabra, no solo con color.
              ChipEstado(
                texto: _textoDeEstado(fila.estado),
                tono: _tonoDe(fila.estado),
              ),
            ],
          ),
          const SizedBox(height: Espacio.s1),
          Text(
            _meta(tutorial),
            style: Tipo.cuerpoChico.copyWith(color: t.text3),
          ),
          const SizedBox(height: Espacio.s2),
          Text(
            tutorial.descripcion,
            style: Tipo.cuerpo.copyWith(color: t.text2),
          ),
          if (fila.requisitosPendientes.isNotEmpty) ...[
            const SizedBox(height: Espacio.s2),
            Text(
              '${TextosSoporte.requisitos}: ${fila.requisitosPendientes.map((r) => r.titulo).join(', ')}',
              style: Tipo.cuerpoChico.copyWith(color: t.avisoTexto),
            ),
          ],
          if (fila.fraccion > 0 &&
              fila.estado != EstadoDeProgreso.completado) ...[
            const SizedBox(height: Espacio.s3),
            Progreso(
              valor: fila.fraccion,
              etiqueta: 'Avance de ${tutorial.titulo}',
            ),
          ],
          const SizedBox(height: Espacio.s3),
          Row(
            children: [
              Expanded(
                child: Boton(
                  texto: fila.continuable
                      ? TextosSoporte.continuar
                      : fila.estado == EstadoDeProgreso.completado
                      ? TextosSoporte.repetir
                      : TextosSoporte.comenzar,
                  variante: BotonVariante.primario,
                  onPressed: fila.continuable ? alContinuar : alComenzar,
                ),
              ),
              if (fila.progreso != null) ...[
                const SizedBox(width: Espacio.s2),
                Boton(
                  texto: TextosSoporte.reiniciar,
                  variante: BotonVariante.fantasma,
                  tamano: BotonTamano.sm,
                  onPressed: alReiniciar,
                ),
              ],
            ],
          ),
        ],
      ),
    );
  }

  String _meta(TutorialDefinicion t) {
    final partes = [
      t.categoria,
      _textoDeDificultad(t.dificultad),
      TextosSoporte.pasos(t.pasos.length),
      if (t.minutos != null) TextosSoporte.minutos(t.minutos!),
      if (t.obligatorio) TextosSoporte.obligatorio,
    ];
    return partes.join(' · ');
  }
}

String _textoDeEstado(EstadoDeProgreso estado) => switch (estado) {
  EstadoDeProgreso.pendiente => 'Pendiente',
  EstadoDeProgreso.enProgreso => 'En progreso',
  EstadoDeProgreso.completado => 'Completado',
  EstadoDeProgreso.omitido => 'A medias',
};

String _textoDeDificultad(Dificultad d) => switch (d) {
  Dificultad.inicial => 'Inicial',
  Dificultad.intermedio => 'Intermedio',
  Dificultad.avanzado => 'Avanzado',
};

Tono _tonoDe(EstadoDeProgreso estado) => switch (estado) {
  EstadoDeProgreso.completado => Tono.ok,
  EstadoDeProgreso.enProgreso => Tono.info,
  EstadoDeProgreso.omitido => Tono.aviso,
  EstadoDeProgreso.pendiente => Tono.neutro,
};
